DOC/scheduler.1
.\" Manpage for scheduler.
.TH man 8 "18 Sep 2013" "1.0" "scheduler shell man page"
.SH NAME
scheduler \- command line interface to scheduler.pl process.
.SH SYNOPSIS
$ .../CODE/shells/scheduler <command>
.SH COMMANDS
One of { start | pause | resume | status | verbose | quiet |
         qs | jobq | runq | ps | flog | enq | deq |
         submit | killjob | stop | terminate | kill }

\fBstart\fP  : starts the scheduler's (scheduler.pl) process in background
         if not already started.

\fBpause\fP  : temporarily prevents scheduler from executing jobs

\fBresume\fP : resumes operations after \fIpause\fP

\fBstatus\fP : queries the scheduler's status

\fBverbose\fP : turn verbose mode on

\fBquiet\fP   : turn verbose mode off

\fBqs\fP      : display summary internal queues

\fBjobq\fP    : list jobs in queue

\fBrunq\fP    : list running jobs

\fBps\fP      : display executing processes (active kids trees)

\fBflog\fP    : backup and cleanup the log file immediately

\fBenq\fP <resource>  : enq (lock) scheduler's resource(s)

\fBdeq\fP <resources> : deq (free) scheduler's resource(s)

\fBsubmit\fP : submit a job for immediate execution, as specified by an
         additional argument:
         - \fIjid=<N>\fP to request execution of job's <N>
           (as defined in 'jobs' table)
         - an inline job definition of the form:
           \fIXEQ1:'launch text',XEQ2:'routine text',XEQ3:'a1 a2',....\fP

\fBkilljob\fP kid=<pid> : kill the job process of PID <pid> (see \fIrunq\fP)
                    Note that job outputs and temporary files might
                    remain and must be checked

\fBstop\fP : gracefully stops the scheduler's (scheduler.pl) process: it will
       wait in the background for jobs to finish before exiting

\fBterminate\fP : request immediate exit, killing any running jobs

\fBkill\fP : 'stop' didn't do any good ? try brute force 'kill'
       (not recommended)

.SH OPTIONS
n/a
.SH SEE ALSO
perldoc <ROOT_SITE>/CODE/cgi-bin/scheduler.pl
.SH AUTHOR
Didier Lafon
.SH COPYRIGHT
Webobs - 2012-2020 - Institut de Physique du Globe Paris

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.



release-notes.md
# WEBOBS RELEASE NOTES

## Important note
This document contains install/upgrade summary and specific instructions for users and administrators.

The latest release has many improvements, new features and bug fixes.
**Upgrade is recommended for all WebObs administrators**. For known issues, please take a look to [github.com/IPGP/webobs/issues](https://github.com/IPGP/webobs/issues) and do not hesitate to submit any problem with this release.

Sections with `!!` prefix must be carefully read in case of upgrade. It usually means that the upgrade could change some behavior from previous release installations.

## Development version (unrelease yet)

### Enhancements

1. The Scheduler now listens for control commands throught UDP on the `127.0.0.1` local address by default. This will restrict control of the Scheduler to local users only. If you really want to allow other users on the network to access this interface, set the `LISTEN_ADDR` configuration variable to another configured IP address or corresponding hostname in the Scheduler configuration file (by default `scheduler.rc`). In this case, it is advised to limit access to specific hosts using local firewall rules.

2. A new option `MERGE_JOB_LOGS` in the scheduler configuration file (by default `scheduler.rc`) allows to merge the standard output and standard error outputs of the jobs to a common file using the name set in the `logpath` field of the job definition, suffixed by `.log`. If `MERGE_JOB_LOGS` is `y`, `yes`, (case insensitive) or `1`, the output will be merged, otherwise the scheduler follows the historical behaviour and will write the outputs to files respectively suffixed by `.stdout` and `.stderr`.

3. The `setup` script now checks if your system uses the system and service manager _systemd_, and will propose to install or update the service definition files `woscheduler.service` and `wopostboard.service` if necessary, adapting the provided templates to your environment as needed.  The templates files for these systemd services have also been updated.  For further details about WebOBs _systemd_ integration, please refer to the new section `4.4 Postboard and Scheduler system integration` of the WebObs manual.

4. The HTML interface in the Scheduler Manager now uses a checkbox to activate/desactivate a job in the _validity_ field of the job definition form. If you experience some abnormal behaviour (e.g. if the 'validity' checkbox isn't checked when it should), try clearing your browser cache (the javascript and CSS files might not correctly be refreshed).

5. Some bug and security issues were fixed in the Scheduler Runs and Scheduler Manager CGI scripts. It is advised to update your WebObs installation with this release.

### Fixed issues

1. For users using _systemd_, the commands `systemctl stop woscheduler` and `systemctl stop wopostboard` no longer mark the related service as _failed_ as `postboard.pl` and `scheduler.pl` now report a clean exit when stopped through a TERM signal. (Note that the command `sudo systemctl stop woscheduler` will cleanly stop the Scheduler, but running jobs will be killed. If you want to stop the Scheduler after waiting for all the jobs to finish, use the command `/opt/webobs/CODE/shell/scheduler stop`. The systemd service will stop normally when the process exits.)

2. The `scheduler` helper script was not able to start or control the Scheduler on systems where the `netstat` command is missing (on newer systems it is by default replaced by `ss` from the `iproute` toolbox).

3. The postboard and its helper function `WebObs::Config::notify` were rejecting notifications having an empty 4th field (`message` field), although action notifications should rightfully be accepted with no parameters, and therefore with an empty `message` field.

4. When specifying the email address of a webobs user as Envelope From address in a mailing notification submitted to the postboard, the full name of the user and its email address were mistakenly inverted, using something like `john.doe@example.com <John Doe>` instead of `John Doe <john.doe@example.com>` as the `From:` field.


## v2.1.5b (August 2020)

### Enhancements

1. Grid's configuration uses a VIM editor with syntax highlighting. New WEBOBS.rc keys:

```
JS_EDITOR_EDIT_THEME|default
JS_EDITOR_BROWSING_THEME|neat
JS_EDITOR_AUTO_VIM_MODE|no
```

2. Sefran now uses full functionality of `DATASOURCE` variable (Seedlink, Arclink, FDSNWS)

3. MC can handle multiple associated images for each event (still undocumented)

4. GNSS modelling has additional node exclusion filter (distance from target, if defined by `GNSS_TARGET_LATLON`), and fixed scale capability for vectors plot. New keys are:

```
MODELLING_EXCLUDED_FROM_TARGET_KM|
MODELLING_VMAX_MM|0
MODELLING_VMAX_RATIO|.25
MODELLING_VECTORS_CLIP|NO
```

5. Proc request form now includes a "verbose logs" option (debug mode)

6. sefran3.pl now accepts a 'hideloc' CGI parameter similar to the 'hideloc'
   parameter of mc3.pl. If its value is 1, event locations will not be shown in
   the popovers of events in the hourly thumbnail view (which allows the page
   to load faster, especially when using FDSNWS). If not present, its value
   defaults to the inverted value of `MC3_EVENT_DISPLAY_LOC` from the Sefran
   configuration.

7. The MC now includes a calculation of energy in joules, shown for each event
   in a popover in the 'Magnitude' column. The total energy of selected events
   is also shown in the top section of the MC, and two new graphs are
   available: one shows the energy in joules featuring each type of event, the
   other only plots the total energy of all event types.

8. Several improvements in the form EAUX: the list of nodes now uses a fixed
   order, and the entry filters are kept after adding/modifying a new entry to
   ease the entry/modification of new data.


### Fixed issues

1. all procs: cannot change default colormap (Matlab and .cpt file) with `*_COLORMAP` keys


CODE/js/scheduler.js

var plot;
var fontsize = 11;
var options;
var yticks;
var data=new Array();
var TZoffset = -(new Date()).getTimezoneOffset()*60*1000;

function plotRuns() {
	var placeholder = $("#placeholder");
	options = {
		grid: {
			//backgroundColor: { colors: ["#999", "#fff"] },
			backgroundColor: "white" ,
			borderWidth: 1,
			borderColor: "#DDD"
		},
		series: { 
			lines: { show: true, color: "#A59A8D", lineWidth: 8 },
			hoverable: true,
			shadowSize: 0
		},
		xaxis: {
			mode: "time",
			timezone: "browser",
			timeformat: "%y/%m/%d<br>%H:%M",
			minTickSize: [1, "hour"]
		},
		yaxis: {
			ticks: [],
			font: {
				size: 10, lineHeight: 12
			},
			color: "#6B6055",
			min: 0
		},
		selection: {
			mode: "x",
			color: "#D1B999"
		},
		canvas: true
	};
	setData();
	placeholder.bind("plotselected", function (event, ranges) {
		plot = $.plot(placeholder, data, $.extend(true, {}, options, {
			xaxis: {
				 min: ranges.xaxis.from,
				 max: ranges.xaxis.to
			 },
			 canvas: true
		}));
		cop();
	});

	setsizes();
	plot = $.plot(placeholder, data, options);
	cop();
};

$(document).ready(function() {
	// Only run plotRuns() if we are in schedulerRuns.pl, as we are
	// also called from schedulerMgr.pl where it generates an error.
	if ($('#placeholder').length == 1) {
		plotRuns();
	}
});

function plotall() {
	plot.setSelection({ xaxis: { from: options.xaxis.min, to: options.xaxis.max } },true);
	setsizes();
	plot = $.plot(placeholder, data, options);
	cop();
}

function cop(color) {
	var bgcolor = (typeof color == "undefined") ? "#fff" : color ;
	var link = $("#tlsavelink");
	link.hidden;
	try {
		var canvas = plot.getCanvas();
		var context=canvas.getContext("2d");
		context.globalCompositeOperation = "destination-over";
		context.fillStyle = "#fff";
		context.fillRect(0,0,canvas.width,canvas.height);
		var canvasimg = canvas.toDataURL();
		link.attr('href',canvasimg);
		link.attr('download','WebObsSchedTimeLine.png');
		link.show;
	} catch(e) {
		console.log("canvas op failed: "+e);
	}
}

function setsizes() {
	options.yaxis.max = yticks+1;
	options.yaxis.font.size = fontsize;
	options.yaxis.font.lineHeight = fontsize+10;
	options.series.lines.lineWidth = fontsize-4;
	options.xaxis.minTickSize = [1, "hour"];
	if (options.xaxis.max-options.xaxis.max <= 4*86400) {options.xaxis.minTickSize = [10,"minute"]; }
	$('.timeline-container').height((yticks+1)*options.yaxis.font.lineHeight);
}

function positionPopup(){
	$("#ovly").css('display','block');
	$("#overlay_form").css({
		//left: ($(window).width() - $('#overlay_form').width()) / 2,
		//top: ($(window).height() - $('#overlay_form').height()) / 2,
		//position:'absolute'
		left: (($(window).width() - $('#overlay_form').outerWidth()) / 2)+$(window).scrollLeft() + "px",
		position:'fixed'
	});
}
function openPopup(jix) {
	// jix is always required; jix  ==  jobsdefs' html-table row id OR -1 for a new job
	if (arguments.length <= 0 ) { return; } // noop if no jix
	//if (!$("#overlay_form").is(':visible')) { return; } 
	if (jix != -1) { // editing an existing job: populate popup from its table row TDs
		// ATT:  $("td",job.jix)[n] = n (0-based) must match jobsdefs <td> order in def-row  
		$('[name=validity]').prop("checked", $("td", jix)[4].textContent.match(/^Y/g) ? "checked" : "");
		$('[name=res]').val( $("td",jix)[5].textContent);
		$('[name=xeq1]').val( $("td",jix)[6].textContent);
		$('[name=xeq2]').val( $("td",jix)[7].textContent);
		$('[name=xeq3]').val( $("td",jix)[8].textContent);
		$('[name=runinterval]').val( $("td",jix)[9].textContent);
		$('[name=maxsysload]').val( $("td",jix)[10].textContent);
		$('[name=logpath]').val( $("td",jix)[11].textContent);
		$('[name=newjid]').val( $("td",jix)[3].textContent);
		$('[name=jid]').val( $("td",jix)[3].textContent );
		$('[name=action]').val("update");
		$('[name=sendbutton]').attr('onclick',"sendPopup(); return false");

	} else { // inserting a new job
		// DON'T rely on default values for popup defined at html-page creation time
		$('[name=validity]').prop("checked", "");
		$('[name=res]').val("");
		$('[name=xeq1]').val("$WEBOBS{JOB_MCC} genplot");
		$('[name=xeq2]').val("");
		$('[name=xeq3]').val("");
		$('[name=runinterval]').val("3600");
		$('[name=maxsysload]').val("0.8");
		$('[name=logpath]').val("");
		$('[name=jid]').val("-1");
		$('[name=newjid]').val("");
		$('[name=action]').val("insert");
		$('[name=sendbutton]').attr('onclick',"sendPopup(); return false");
	}
	$("#overlay_form").fadeIn(500);
	positionPopup();
}
function closePopup() {
	$("#overlay_form").fadeOut(500);
	$("#ovly").fadeOut(500);
}
function sendPopup() {
	if ( !($('[name=newjid]').val()).match(/^[a-zA-Z0-9\-_]+$/) ) {
		alert ("required jid, [a-zA-Z0-9\-_] only)");
		return false;
	}
	if ( !($('[name=runinterval]').val()).match(/^[0-9]+$/) ) {
		alert ("runInterval must be numeric");
		return false;
	}	
	if ( !($('[name=maxsysload]').val()).match(/^[0-9.]+$/) ) {
		alert ("maxsysload must be a float");
		return false;
	}
	if ( $('[name=logpath]').val() == "" ) {
		alert ("logpath can't be empty");
		return false;
	}
	if ( $('[name=xeq1]').val() == "" && $('[name=xeq2]').val() == "" && $('[name=xeq3]').val() == "" ) {
		alert ("invalid: nothing to be executed!");
		return false;
	}
	if ( $('[name=jid]').val() == "-1"  ) {
		$('[name=jid]').val($('[name=newjid]').val());
	}
	postInsUpd();
}
function postInsUpd() {
	$("#overlay_form").fadeOut(500);
	$("#ovly").fadeOut(500);
	location.href = "/cgi-bin/schedulerMgr.pl?"+$("#overlay_form").serialize();
}
function postDelete(jix) {
	var jid =  $("td",jix)[3].textContent;
	var answer = confirm("do you really want to delete jid "+jid+ " ?")
	if (answer) {
		location.href = "/cgi-bin/schedulerMgr.pl?action=delete&jid="+jid
	}
}
function postSubmit(jix) {
	var jid =  $("td",jix)[3].textContent;
	var answer = confirm("do you really want to submit jid "+jid+ " ?")
	if (answer) {
		location.href = "/cgi-bin/schedulerMgr.pl?action=submit&jid="+jid
	}
}
function postKill(jix) {
	var answer = confirm("do you really want to kill job process kid "+jix+" ?")
	if (answer) {
		location.href = "/cgi-bin/schedulerRuns.pl?action=killjob&kid="+jix
	}
}

//---   $("form").serialize() 
//$("td",job.jid).each(function() { 
//	console.log(this.textContent) ;
//});
